src/rust/engine/src/context.rs

```diff
@@ -121,9 +122,22 @@ impl Core {
   where
     F: Fn() -> T,
   {
-    self.fs_pool.with_shutdown(|| {
-      self.runtime.with_reset(|| {
-        self.graph.with_exclusive(|| {
+    // Only one fork may occur at a time, but draining the Runtime and Graph requires that the
+    // Graph lock is not actually held during draining (as that would not allow the Runtime's
+    // threads to observe the draining value). So we attempt to mark the Graph draining (similar
+    // to a CAS loop), and treat a successful attempt as indication that our thread has permission
+    // to execute the fork.
+    //
+    // An alternative would be to have two locks in the Graph: one outer lock for the draining
+    // bool, and one inner lock for Graph mutations. But forks should be rare enough that busy
+    // waiting is not too contentious.
+    while let Err(()) = self.graph.mark_draining(true) {
+      debug!("Waiting to enter fork_context...");
+      thread::sleep(Duration::from_millis(10));
+    }
+    let t = self.runtime.with_reset(|| {
```

Comment:
We should really switch `Resettable.get` to `Resettable.with` at some point... May make a good cross-pollination ticket, actually, for getting someone into Rust :)